pkg/provisioner/ironic/ironic.go

```diff
@@ -319,13 +341,73 @@ func (p *ironicProvisioner) changeNodeProvisionState(ironicNode *nodes.Node, opt
 	return result, nil
 }
 
+func getNICDetails(ifdata []introspection.InterfaceType) []metalkubev1alpha1.NIC {
+	nics := make([]metalkubev1alpha1.NIC, len(ifdata))
+	for i, intf := range ifdata {
+		nics[i] = metalkubev1alpha1.NIC{
+			Name:      intf.Name,
+			Model:     fmt.Sprintf("%s %s", intf.Vendor, intf.Product),
+			MAC:       intf.MACAddress,
+			Network:   "Pod Networking", // TODO(zaneb)
+			IP:        intf.IPV4Address,
+			SpeedGbps: 0, // TODO(zaneb)
+		}
+	}
+	return nics
+}
+
+func getStorageDetails(diskdata []introspection.RootDiskType) []metalkubev1alpha1.Storage {
+	storage := make([]metalkubev1alpha1.Storage, len(diskdata))
+	for i, disk := range diskdata {
+		storage[i] = metalkubev1alpha1.Storage{
+			Name:    disk.Name,
+			Type:    map[bool]string{true: "HDD", false: "SSD"}[disk.Rotational],
+			SizeGiB: metalkubev1alpha1.GiB(disk.Size / (1024 * 1024 * 1024)),
+			Model:   fmt.Sprintf("%s %s", disk.Vendor, disk.Model),
+		}
+	}
+	return storage
+}
+
+func getCPUDetails(cpudata *introspection.CPUType) []metalkubev1alpha1.CPU {
+	var freq float64
+	fmt.Sscanf(cpudata.Frequency, "%fGHz", &freq)
+	cpu := metalkubev1alpha1.CPU{
+		Type:     cpudata.Architecture,
+		SpeedGHz: metalkubev1alpha1.GHz(freq),
+	}
+
+	cpus := make([]metalkubev1alpha1.CPU, cpudata.Count)
+	for i := range cpus {
+		cpus[i] = cpu
+	}
+	return cpus
+}
+
+func getHardwareDetails(data *introspection.Data) *metalkubev1alpha1.HardwareDetails {
+	details := new(metalkubev1alpha1.HardwareDetails)
+	details.RAMGiB = metalkubev1alpha1.GiB(data.MemoryMB / 1024)
+	details.NIC = getNICDetails(data.Inventory.Interfaces)
+	details.Storage = getStorageDetails(data.Inventory.Disks)
+	details.CPUs = getCPUDetails(&data.Inventory.CPU)
+	return details
+}
+
 // InspectHardware updates the HardwareDetails field of the host with
 // details of devices discovered on the hardware. It may be called
 // multiple times, and should return true for its dirty flag until the
 // inspection is completed.
 func (p *ironicProvisioner) InspectHardware() (result provisioner.Result, err error) {
 	p.log.Info("inspecting hardware", "status", p.host.OperationalStatus())
 
+	// Introspection not started
+	if p.host.Status.Provisioning.State != metalkubev1alpha1.StateInspecting {
```

Comment:
I don't disagree that it should, but nothing in #121 attempts to do that AFAICT. That seems like something to do in a separate PR?

Reply:
Maybe I pointed to the wrong PR. I don't think ironic.go sets Provisioning.State anywhere any more, so I was suggesting we not add a new place where it does.

Reply:
It was still being set on line 368, this just moves it to line 351.

Reply:
OK, we can fix that in another PR then

Reply:
Actually I'm wrong, it was added in the previous patch (at least after rebasing). I see now where it's getting set in #121 (it wasn't where I expected, and I think I thought it wasn't happening because it no longer records an event). I'll squash those two patches and make sure we're doing all state changes only in the controller.